### PR TITLE
expression, util/codec: fix wrongly eliminated conditions caused by `HashCode()` collision

### LINKED ISSUE #30094: Where conditions are eliminated wrongly caused by `HashCode` collision
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```sql
create table t(a varchar(10));
explain select * from t where cast(a as float) and cast(a as char);
explain select * from t where a = 0xe59388e59388e59388 and a = 0xe598bfe598bfe598bf;
```

UPDATE: the second SQL can't trigger the bug anymore. Please switch to use the following SQL:
```sql
explain select * from t30094 t1 where  concat(t1.a,'1') = _binary 0xe59388e59388e59388 collate binary and concat(t1.a,'1') = _binary 0xe598bfe598bfe598bf collate binary;
```

### 2. What did you expect to see? (Required)

The selection operator has two conditions.

### 3. What did you see instead (Required)

```sql
> explain select * from t where cast(a as float) and cast(a as char);
+-------------------------+----------+-----------+---------------+--------------------------------+
| id                      | estRows  | task      | access object | operator info                  |
+-------------------------+----------+-----------+---------------+--------------------------------+
| TableReader_7           | 8000.00  | root      |               | data:Selection_6               |
| └─Selection_6           | 8000.00  | cop[tikv] |               | cast(test2.t.a, float BINARY)  |
|   └─TableFullScan_5     | 10000.00 | cop[tikv] | table:t       | keep order:false, stats:pseudo |
+-------------------------+----------+-----------+---------------+--------------------------------+

> explain select * from t where a = 0xe59388e59388e59388 and a = 0xe598bfe598bfe598bf;
+-------------------------+----------+-----------+---------------+---------------------------------------+
| id                      | estRows  | task      | access object | operator info                         |
+-------------------------+----------+-----------+---------------+---------------------------------------+
| TableReader_7           | 10.00    | root      |               | data:Selection_6                      |
| └─Selection_6           | 10.00    | cop[tikv] |               | eq(test2.t.a, "0xe59388e59388e59388") |
|   └─TableFullScan_5     | 10000.00 | cop[tikv] | table:t       | keep order:false, stats:pseudo        |
+-------------------------+----------+-----------+---------------+---------------------------------------+
```

### 4. What is your TiDB version? (Required)

master

### PR DESCRIPTION
**Pull request by time-and-fate**
### What problem does this PR solve?

Issue Number: close #30094

Problem Summary:

Sometimes (please see the issue), the `HashCode` of expressions may collide, then the expressions are wrongly eliminated in `RemoveDupExprs()`.

The collisions are caused by:
1. `Cast` is a special function, where its `RetType` should also be considered as an argument. But we didn't handle this before.
2. `BinaryLiteral` might overflow. In this case, all overflowed `BinaryLiteral` will become a `MaxUint64`

### What is changed and how it works?

In the `HashCode()`:
1. Add additional logic for the `Cast`.
2. Use a new lossless encoding method to replace `EncodeValue`.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Review by Reminiscent — APPROVED** [review]

File: util/codec/codec.go [hunk 1]
```diff
@@ -1294,3 +1296,57 @@ func ConvertByCollationStr(str string, tp *types.FieldType) string {
...
+	case types.KindFloat32, types.KindFloat64:
+		b = append(b, floatFlag)
+		b = EncodeFloat(b, d.GetFloat64())
+	case types.KindString:
+		b = encodeString(b, d, false)
+	case types.KindBytes:
+		b = encodeBytes(b, d.GetBytes(), false)
+	case types.KindMysqlTime:
+		b = append(b, uintFlag)
+		t := d.GetMysqlTime().CoreTime()
```
Comment (AilinKid) [inline util/codec/codec.go:1318]: need timezone here?

Comment (AilinKid) [reply util/codec/codec.go:1318]: seems ok if hashcode only used to deduplicate expr in one plan

**Review by AilinKid — APPROVED** [review]
LGTM

File: util/codec/codec.go (hunk 1 above)
Comment (time-and-fate) [reply util/codec/codec.go:1318]: Yes. It's only used in the planner and it's in the same session. I think it's OK.

**Pull request merged**